Umee-Bots/Large-Borrow/src/utils.ts

```diff
@@ -0,0 +1,112 @@
+import { Mutex } from "async-mutex";
+import BigNumber from "bignumber.js";
+import { CallOverrides } from "ethers";
+import { ethers, Finding, FindingSeverity, FindingType } from "forta-agent";
+import LRU from "lru-cache";
+
+export interface AgentConfig {
+  lendingPoolAddress: string;
+  tvlPercentageThreshold: string;
+}
+
+export function createFinding(
+  amount: ethers.BigNumber,
+  tvlPercentage: BigNumber,
+  user: string,
+  onBehalfOf: string
+): Finding {
+  return Finding.from({
+    alertId: "UMEE-6",
+    name: "Large borrow",
+    description: "There was a large borrow based on the pool's TVL",
+    type: FindingType.Info,
+    severity: FindingSeverity.Unknown,
+    protocol: "Umee",
+    metadata: {
+      amount: amount.toString(),
+      tvlPercentage: tvlPercentage.toString(10),
+      user: user.toLowerCase(),
+      onBehalfOf: onBehalfOf.toLowerCase(),
+    },
+  });
+}
+
+export interface SmartCallerOptions {
+  cacheByBlockTag: boolean;
+}
+
+export class SmartCaller {
+  static cache = new LRU<string, Promise<any>>({ max: 200 });
+  static mutex = new Mutex();
+
+  [key: string]: ((...args: any[]) => Promise<any>) | any;
+  contract: ethers.Contract;
+  options: SmartCallerOptions;
+
+  constructor(contract: ethers.Contract, options: Partial<SmartCallerOptions> = {}) {
+    this.contract = contract;
+    this.options = {
+      cacheByBlockTag: true,
+      ...options,
+    };
+
+    this.generateMethods();
+  }
+
+  static from(contract: ethers.Contract, options: Partial<SmartCallerOptions> = {}): SmartCaller {
+    return new SmartCaller(contract, options);
+  }
+
+  protected generateMethods() {
+    Object.keys(this.contract.functions).forEach((functionName) => {
+      const inputsLength = this.contract.interface.getFunction(functionName).inputs.length;
+
+      this[functionName] = (...args: any[]): Promise<any> => {
+        let overrides = {};
+        if (args.length === inputsLength + 1 && typeof args[args.length - 1] === "object") {
+          Object.assign(overrides, args.pop());
+        }
+
+        return SmartCaller.smartCall(this.contract, functionName, args, overrides, this.options);
+      };
+    });
+  }
+
+  protected static async smartCall(
+    contract: ethers.Contract,
+    functionName: string,
+    args: any[],
+    overrides: CallOverrides = {},
+    options: SmartCallerOptions
```

Comment:
this should not work :thinking:, before a default param all the other params needs to have a default value too

Reply:
Yes, I added `options` after `overrides` and didn't notice that - but it does work 🤔 . Changed now, thanks!